third_party/ibis/ibis_postgres/client.py

```diff
@@ -40,72 +49,106 @@ def _get_schema_using_query(self, query: str) -> sch.Schema:
     tuples = [(col, self._get_type(typestr)) for col, typestr in type_info]
     return sch.Schema.from_tuples(tuples)
 
-def _get_type(self,typestr: str) -> dt.DataType:
-    _type_mapping = {
-    # "boolean": dt.bool,
+
+_BRACKETS = "[]"
+_STRING_REGEX = (
+    """('[^\n'\\\\]*(?:\\\\.[^\n'\\\\]*)*'|"[^\n"\\\\"]*(?:\\\\.[^\n"\\\\]*)*")"""
+)
+
+
+def spaceless(parser):
+    return SPACES.then(parser).skip(SPACES)
+
+
+def spaceless_string(*strings: str):
+    return spaceless(
+        parsy.alt(*(parsy.string(s, transform=str.lower) for s in strings))
+    )
+
+
+SPACES = parsy.regex(r"\s*", re.MULTILINE)
+RAW_NUMBER = parsy.decimal_digit.at_least(1).concat()
+SINGLE_DIGIT = parsy.decimal_digit
+PRECISION = SCALE = NUMBER = RAW_NUMBER.map(int)
+
+LPAREN = spaceless_string("(")
+RPAREN = spaceless_string(")")
+
+LBRACKET = spaceless_string("[")
+RBRACKET = spaceless_string("]")
+
+LANGLE = spaceless_string("<")
+RANGLE = spaceless_string(">")
+
+COMMA = spaceless_string(",")
+COLON = spaceless_string(":")
+SEMICOLON = spaceless_string(";")
+
+RAW_STRING = parsy.regex(_STRING_REGEX).map(ast.literal_eval)
+FIELD = parsy.regex("[a-zA-Z_][a-zA-Z_0-9]*")
+
+
+def _parse_numeric(
```

Comment:
This feels over-engineered to me. I looked at how the Impala client does this and it uses `dt.dtype` like this:
```
        elif "decimal" in t:
            result = dt.dtype(t)
```

If we followed that approach for PostgreSQL your new `_get_type` function could look like this:
```
def _get_type(self, typestr: str) -> dt.DataType:
    typestr = typestr.lower()
    is_array = typestr.endswith(_BRACKETS)
    typ = _type_mapping.get(typestr.replace(_BRACKETS, ""))
    if typ is not None:
        return dt.Array(typ) if is_array else typ
    if typestr.startswith("numeric"):
        # numeric is a synonym for decimal
        typestr = typestr.replace("numeric", "decimal")
    if typestr.lower().startswith("decimal"):
        return dt.dtype(typestr)
    return _type_mapping[typestr]
```

The code is bit more primitive, no AST etc, but it means you don't need `parsy`,  `_parse_numeric` or any of the new constants/imports.

I think the intention would be a lot clearer.

Did I miss something?

Reply:
Can you provide the table/query you used?

We have integration test written already for major datatypes and it works for them. I have checked it as well.

**table creation:**
DROP TABLE IF EXISTS test_data_types_postgres_row;
CREATE TABLE test_data_types_postgres_row(
	serial_col serial PRIMARY KEY,
    int_col int,
	text_col  text,
    char_col  char,
    varchar_col  varchar,
	float_col double precision,
	numeric_col  numeric(5, 2),
	timestamp_col timestamp,
    date_col date
);
INSERT INTO test_data_types_postgres_row(int_col, text_col, char_col, varchar_col, float_col, numeric_col, timestamp_col, date_col) values (10, 'row1_text', '1', 'row1_varchar', 10.10, 10.101, '2016-06-22 19:10:25-07','2020-01-01');
INSERT INTO test_data_types_postgres_row(int_col, text_col, char_col, varchar_col, float_col, numeric_col, timestamp_col, date_col) values (20, 'row2_text', '2', 'row2_varchar', 20.20, 20.202, '2016-06-22 19:20:25-07','2020-02-02');
INSERT INTO test_data_types_postgres_row(int_col, text_col, char_col, varchar_col, float_col, numeric_col, timestamp_col, date_col) values (30, 'row3_text', '3', 'row3_varchar', 30.20, 30.203, '2016-06-22 19:30:25-07','2020-03-03');
INSERT INTO test_data_types_postgres_row(int_col, text_col, char_col, varchar_col, float_col, numeric_col, timestamp_col, date_col) values (40, 'row4_text', '4', 'row4_varchar', 40.20, 40.404, '2016-06-22 19:30:45-07','2020-04-04');
INSERT INTO test_data_types_postgres_row(int_col, text_col, char_col, varchar_col, float_col, numeric_col, timestamp_col, date_col) values (50, 'row5_text', '5', 'row5_varchar', 50.20, 50.404, '2016-06-22 19:50:44-0','2020-05-05');
INSERT INTO test_data_types_postgres_row(int_col, text_col, char_col, varchar_col, float_col, numeric_col, timestamp_col, date_col) values (60, 'row6_text', '6', 'row6_varchar', 60.20, 60.404, '2016-01-22 19:51:44-0','2020-06-06');
INSERT INTO test_data_types_postgres_row(int_col, text_col, char_col, varchar_col, float_col, numeric_col, timestamp_col, date_col) values (70, 'row7_text', '7', 'row7_varchar', 70.20, 70.404, '2017-01-22 19:51:44-0','2020-06-07');
INSERT INTO test_data_types_postgres_row(int_col, text_col, char_col, varchar_col, float_col, numeric_col, timestamp_col, date_col) values (80, 'row8_text', '8', 'row8_varchar', 80.20, 80.404, '2018-01-22 19:51:44-0','2020-06-08');
INSERT INTO test_data_types_postgres_row(int_col, text_col, char_col, varchar_col, float_col, numeric_col, timestamp_col, date_col) values (90, 'row9_text', '9', 'row9_varchar', 90.20, 90.404, '2019-01-22 19:51:44-0','2020-06-09');
INSERT INTO test_data_types_postgres_row(int_col, text_col, char_col, varchar_col, float_col, numeric_col, timestamp_col, date_col) values (100, 'row10_text', '1', 'row10_varchar', 100.20, 100.404, '2020-01-22 19:51:44-0','2021-06-09');

**dvt execution**
data-validation validate custom-query --custom-query-type row --source-query-file source_query.sql --target-query-file source_query.sql -sc MY_PG_CONN -tc MY_PG_CONN --hash '*' --primary-keys serial_col

Reply:
Ok, fair enough regarding the code.